rmqsrc/encode.nim

```diff
@@ -0,0 +1,102 @@
+import endians, tables, strutils
+import spec, frame
+
+# Encode basic types
+proc encode(v: uint8): array[0..0, char] =
+  return [v.char]
+
+proc encode(v: uint16): array[0..1, char] =
+  var v = v
+  bigEndian16(addr result, addr v)
+
+proc encode(v: uint32): array[0..3, char] =
+  var v = v
+  bigEndian32(addr result, addr v)
+
+proc encode(v: string): seq[char] =
+  result = newSeq[char]()
+  for ch in v:
+    result.add(ch)
+
+proc encode(v: Table[string, string]): seq[char] =
+  # TODO fill this out, right now we will just assume the table is empty
+  result = newSeq[char]()
+  for ch in encode(0.uint32):
+    result.add(ch)
+
+# Define encodings method parameters
+type 
+  ConnectionStartOk = tuple
+    index: uint32
+    clientProps: Table[string, string]
+    mechanism: (string, uint8) # short string
+    response: (string, uint32) # long string
+    locale: (string, uint8)    # short string
+
+proc encode*(v: ConnectionStartOk): seq[char] =
+  result = newSeq[char]()
```

Comment:
Since you know how long the sequence is, you could either use an array, or at least pass a length argument to `newSeq`.

Reply:
We can't use an array since we don't know the full length before runtime (due to the strings and tables). Creating a seq with a known length is a good idea tho.

Reply:
Don't you know the full length, though? You call `add` 8 times - isn't it always going to be 8 elements long?